### ISSUE TITLE
SUComponentInstanceSaveAs modifies the original instance and makes it unique

### ISSUE THREAD

**Issue by AnFunctionArray** [SketchUp/api-issue-tracker#344]
The issue is that SUComponentInstanceGetTransform always returns identity when requesting on a parsed component instance from a [model skp file](https://forums.sketchup.com/uploads/short-url/aWMKUyX8PiMCa4IAIGC2CnaCjyn.skp).

Consider the following code:

```cpp
#include <stdio.h>
#include <iostream>
#include <SketchUpAPI/common.h>
#include <SketchUpAPI/geometry.h>
#include <SketchUpAPI/initialize.h>
#include <SketchUpAPI/model/model.h>
#include <SketchUpAPI/model/entities.h>
#include <SketchUpAPI/model/face.h>
#include <SketchUpAPI/model/edge.h>
#include <SketchUpAPI/model/vertex.h>
#include <SketchUpAPI/model/component_definition.h>
#include <SketchUpAPI/model/component_instance.h>
#include <SketchUpAPI/geometry/transformation.h>

int main(int n, char **args) {
	// Always initialize the API before using it
	SUInitialize();

	if (n > 1 && !strcmp(1[args], "display"))
	{
		SUModelRef model = SU_INVALID;

		SUModelCreateFromFile(&model, "new_model.skp");

		SUEntitiesRef entities = SU_INVALID;
		SUModelGetEntities(model, &entities);

		size_t count = 0;

		SUEntitiesGetNumInstances(entities, &count);
		std::cout << "instance count: " << count << "\n";

		SUComponentInstanceRef* instances = (SUComponentInstanceRef*)_alloca(count * sizeof * instances);

		size_t szinstances;

		SUEntitiesGetInstances(entities, count, instances, &szinstances);
		std::cout << "returned instances count: " << szinstances << "\n";
		while (szinstances--)
		{
			struct SUTransformation matrix;
			SUComponentInstanceGetTransform(instances[szinstances], &matrix);
			struct FloatingPosition {
				double x, z, y, d;
			} pos = { .x = matrix.values[12], .z = matrix.values[14], .y = matrix.values[13], .d = matrix.values[15] };
			printf("%f, %f, %f\n", pos.x, pos.y, pos.z);
		}

		return 0;
	}
	// Create an empty model
	SUModelRef model = SU_INVALID;
	SUResult res = SUModelCreate(&model);
	// It's best to always check the return code from each SU function call.
	// Only showing this check once to keep this example short.
	if (res != SU_ERROR_NONE)
		return 1;
	// Get the entity container of the model
	SUEntitiesRef entities = SU_INVALID;
	SUModelGetEntities(model, &entities);
	// Create a loop input describing the vertex ordering for a face's outer loop
	SULoopInputRef outer_loop = SU_INVALID;
	SULoopInputCreate(&outer_loop);
	for (size_t i = 0; i < 4; ++i) {
		SULoopInputAddVertexIndex(outer_loop, i);
	}
	// Create the face
	SUFaceRef face = SU_INVALID;
	SUPoint3D vertices[4] = { { 0,   0,   0 },
							  { 100, 100, 0 },
							  { 100, 100, 100 },
							  { 0,   0,   100 } };
	SUFaceCreate(&face, vertices, &outer_loop);
	// Add the face to the entities
	SUComponentDefinitionRef def = SU_INVALID;
	SUComponentDefinitionCreate(&def);
	SUEntitiesRef defentities = SU_INVALID;
	SUComponentDefinitionGetEntities(def, &defentities);
	SUEntitiesAddFaces(defentities, 1, &face);
	SUComponentInstanceRef instance = SU_INVALID;
	SUComponentDefinitionCreateInstance(def, &instance);
	static SUTransformation matrixtransform;
	constexpr SUVector3D point = { 666, 666, 666 };
	SUTransformationTranslation(&matrixtransform, &point);
	SUComponentInstanceSetTransform(instance, &matrixtransform);
	SUEntitiesAddInstance(entities, instance, nullptr);

	//comment the next line to make it work
	SUComponentInstanceSaveAs(instance, "instance.skp");
	// Save the in-memory model to a file
	res = SUModelSaveToFile(model, "new_model.skp");
	// Must release the model or there will be memory leaks
	SUModelRelease(&model);
	// Always terminate the API when done using it
	SUTerminate();

	if (res != SU_ERROR_NONE) {
		std::cout << "Unable to save document!";
		return 1;
	}

	std::cout << "Document saved!";
	return 0;
}
```

I'm using SDK_WIN_x64_2019-2-222. The above code will always print 0.000000, 0.000000, 0.000000 - component_instance_name - no matter if the component instance have any transformation taken in place (I exported them to skp with a transformation).

**Comment by DanRathbun** [comment]
FYI: Markdown for multiline code is ...

\`\`\`c
`  // code goes here`
\`\`\`


Where the first line delimiter has the lowercase name of the language for lexing.
_See: **[Mastering Markdown: GitHub Flavoured Markdown](https://guides.github.com/features/mastering-markdown)** , scroll down to "**Syntax highlighting**"_

([Note this also works for the Discourse driven SketchUp forums](https://forums.sketchup.com/t/how-to-colourize-code-on-the-forum-wiki/46189).)

**Comment by bsld** [comment]
Here is [new_model.skp](https://forums.sketchup.com/uploads/short-url/aWMKUyX8PiMCa4IAIGC2CnaCjyn.skp) on which I'm trying the above program with a bunch of zeroes as output.

**Comment by DanRathbun** [comment]
As posted here ...
https://forums.sketchup.com/t/getting-transformation-from-an-instance-obtained-from-a-model-always-return-identity/105131/6

... it is the model's definitions that were not created correctly.

Basically it was the definition's entities that were transformed instead of applying the transformation to the instances. This also created the scenario where there are unique definition's for instances that are alike. (Ie, there should be far less definition's in the file.)

![image](https://user-images.githubusercontent.com/334837/66258649-c98a8880-e775-11e9-9562-eef00eebafcf.png)

**Comment by bsld** [comment]
Here is my [code](https://github.com/bsld/RMXtoskp). On line [250@zone.c](https://github.com/bsld/RMXtoskp/blob/master/zone.c#L250) I clearly specify `SUComponentInstanceSetTransform`.

_The input [files](https://forums.sketchup.com/uploads/short-url/mycbNXXcxnvuzWOzP8NQ4bSy4YT.zip) I use._

**Comment by bsld** [comment]
Aha I found the bug - it's the fact that I'm saving individual components on line [256@zone.c](https://github.com/bsld/RMXtoskp/blob/master/zone.c#L256) - if you comment said line it's showing the transformations back in the reader.

**Comment by thomthom** [comment]
What's the current status of this? Is there still a claim for a bug in the SketchUp API? (If so, can the original post be updated to include a complete example reproducing it?)

(Btw, @bsld, your example leak memory in the loop. When you use `SUStringCreate` you also need to use `SUStringRelease`.)

**Comment by bsld** [comment]
@thomthom I updated the OP with example showing the bug - there is a comment above the inflicting line.

**Comment by DanRathbun** [comment]
@thomthom I think that using the `SUComponentInstanceSaveAs(instance, "instance.skp");` statement works like the **`Make Component`** command in the application. It makes a unique definition and **replaces** the instance with an new instance of this uniqueified definition (hence why the "In Model" definition list is so bloated with cloned components.) These new instances have their entities origin at the model's origin as a default it seems, which displaces their geometry. (Take a look at the model he posted.)

**Comment by thomthom** [comment]
Ugh - yea. `SUComponentInstanceSaveAs` is doing some strange things. Baking the instance transformation into the definition probably (??) makes sense for the exported model. But it shouldn't modify the original like it currently does.

I'm also surprised that there isn't any `SUComponentDefinitionSaveAs`.

Until we can backfill such functionality, can you work around this by creating a temporary instance that you save?

**Comment by DanRathbun** [comment]
> @thomthom Ugh - yea. `SUComponentInstanceSaveAs` is doing some strange things. Baking the instance transformation into the definition probably (??) makes sense for the exported model. But it shouldn't modify the original like it currently does.

"Strange things" is why I questioned what it's documentation meant by "**serialization**" in issue 71 (a year and a half ago) ...
* [C API: SUComponentInstanceSaveAs() function use is unclear](https://github.com/SketchUp/api-issue-tracker/issues/71)

----

P.S. - I also noted in that issue the fact that the C API had the opposite of the Ruby API (save as for the instance instead of the definition.)

**Comment by jinyistyle** [comment]
SKOR-12616

**Comment by DanRathbun** [comment]
@thomthom This issue and Jira # SKOR-12616 can likely be closed as v2022.0 (API v10.0) has added [`SUComponentDefinitionSaveToFile()`](https://extensions.sketchup.com/developers/sketchup_c_api/sketchup/struct_s_u_component_definition_ref.html#a98735078e64bf09c9f6ff04d295b9b88) and deprecated the flawed [`SUComponentInstanceSaveAs()`](https://extensions.sketchup.com/developers/sketchup_c_api/sketchup/struct_s_u_component_instance_ref.html#add89b2be37466aa7499213e5289911cd) function.

@AnFunctionArray Suggest you update to API v10.0 and migrate to using the new [`SUComponentDefinitionSaveToFile()`](https://extensions.sketchup.com/developers/sketchup_c_api/sketchup/struct_s_u_component_definition_ref.html#a98735078e64bf09c9f6ff04d295b9b88) function.
